device/cuda/src/details/sparse_ccl.cu

```diff
@@ -0,0 +1,343 @@
+        );
+    }
+
+    __global__
+    void
+    sparse_ccl_kernel(
+        const cell * _cells,
+        const unsigned int * blocks,
+        float * _out
+    ) {
+        /*
+         * Select the relevant output region for this block. Since we are
+         * writing 4 floats per block (module) and we have a maximum of
+         * MAX_CLUSTERS_PER_MODULE output clusters per block, we need to seek
+         * four times that length into the array for each block.
+         */
+        float * out = &_out[4 * MAX_CLUSTERS_PER_MODULE * blockIdx.x];
+
+        /*
+         * Seek the correct cell region in the input data. Again, this is all a
+         * contiguous block of memory for now, and we use the blocks array to
+         * define the different ranges per block/module. At the end of this we
+         * have the starting address of the block of cells dedicated to this
+         * module, and we have its size.
+         */
+        const cell * cells = &_cells[blocks[blockIdx.x]];
+        const unsigned int size = blocks[blockIdx.x + 1] - blocks[blockIdx.x];
+
+        /*
+         * As an optimisation, we will keep track of which cells are adjacent to
+         * each other cell. To do this, we define, in thread-local memory or
+         * registers, up to eight adjacent cell indices and we keep track of how
+         * many adjacent cells there are (i.e. adjc[i] determines how many of
+         * the eight values in adjv[i] are actually meaningful).
+         *
+         * The implementation is such that a thread might need to process more
+         * than one hit. As such, we keep one counter and eight indices _per_
+         * hit the thread is processing. This number is never larger than
+         * the max number of activations per module divided by the threads per
+         * block.
+         */
+        unsigned int adjv[MAX_ACTIVATIONS_PER_MODULE / THREADS_PER_BLOCK][8];
+        unsigned int adjc[MAX_ACTIVATIONS_PER_MODULE / THREADS_PER_BLOCK];
+
+        /*
+         * This variable will be used to write to the output later. We declare
+         * and set it here because it saves us a synchronisation barrier. Not a
+         * huge deal, but worth keeping in mind.
+         */
+        __shared__ unsigned int outi;
+
+        if (threadIdx.x == 0) {
+            outi = 0;
+        }
+
+        /*
+         * These arrays are the meat of the pudding of this algorithm, and we
+         * will constantly be writing and reading from them which is why we
+         * declare them to be in the fast shared memory. Note that this places a
+         * limit on the maximum activations per module, as the amount of shared
+         * memory is limited. These could always be moved to global memory, but
+         * the algorithm would be decidedly slower in that case.
+         */
+        __shared__ unsigned int f[MAX_ACTIVATIONS_PER_MODULE], gf[MAX_ACTIVATIONS_PER_MODULE];
```

Comment:
You could be a bit more explicit (in the comment or the variables names) what `f` and `gf` **are**.

Reply:
Yes, this is absolutely true. Good point, thank you! :smile: